[exporter/datadog] Add Kubernetes hostname provider

mx-psi opened the pull request:
**Description:** 

- Add Kubernetes hostname provider for the preview hostname logic.
- Fixes some issues on the chain hostname provider (will add more tests on a future PR)

**Link to tracking Issue:** #10424

**Testing:** Added unit tests. TBD: end to end tests

mx-psi committed:
[exporter/datadog] Add Kubernetes hostname provider

dashpole commented on exporter/datadogexporter/internal/metadata/internal/k8s/nodename.go:
```diff
@@ -0,0 +1,63 @@
+		return "", fmt.Errorf("could not fetch our hostname: %w", err)
+	}
+
+	pod, err := p.client.CoreV1().Pods(namespace).Get(ctx, podName, metav1.GetOptions{})
```
should probably document the required rbac rules.

mx-psi replied:
It turns out the RBAC rules we currently document for supporting the k8sattributes processor are sufficient for this, so I just noted in the code that these should be kept in sync with the ones on our docs/examples

dashpole commented on exporter/datadogexporter/internal/metadata/internal/k8s/nodename.go:
```diff
@@ -0,0 +1,63 @@
+
+func (p *nodeNameProviderImpl) NodeName(ctx context.Context) (string, error) {
+	namespace := p.namespace()
+	podName, err := os.Hostname()
```
note: this isn't guaranteed to be the pod name

mx-psi replied:
Is there a better way to do this? I am taking the same approach as the one the Datadog Agent uses, but I trust you more than our codebase 😄

dashpole replied:
Using the downward API would be better, but there isn't other than that.  It might be worth documenting, though, that setting the `hostname` or using `setHostnameAsFQDN` in your pod spec will break detection.

dashpole reviewed (commented):
Why not use resource detectors, and then reference resource attributes in this exporter?  It seems like this is reimplementing most resource detection code.

mx-psi commented:
> Why not use resource detectors, and then reference resource attributes in this exporter? It seems like this is reimplementing most resource detection code.

@dashpole That's a fair question. We have tried to rely as much as possible on the resource detection processor/k8sattributes processor so far (e.g. our current implementation requires you to use the k8sattributes processor if on Kubernetes for certain environments), but with this approach it's not possible to support some setups without passing significant complexity to users. Our exporter needs to send a payload with information about the 'host' in which it is running, but when information like `k8s.node.name` comes in a metrics/traces/logs payload, it's not possible to know if it comes from the host in which the Collector is running or from a remote host.

That said, I am trying to not duplicate code (see #10376) and will do future refactors to reduce code duplication more, so that in that the duplication is reduced as much as possible.

mx-psi committed:
Add missing license header and import comment

mx-psi committed:
Fix linter error

mx-psi committed:
[exporter/datadog] Make rules consistent on preview attributes rules and provider rules

mx-psi committed:
Add notes about pod name and RBAC rules

mx-psi committed:
More linter fixes

mx-psi committed:
[exporter/datadog] Make node name provider be available unconditionally

We can't make it fail at provider build time, since otherwise it will make the whole provider
build fail if we are not on Kubernetes
